File: 2021/src/day_7.rs
use std::fs;

pub const TITLE: &str = "--- Day 7: The Treachery of Whales ---";

pub fn part_1() -> i64 {
    let mut crab_positions = fs::read_to_string("inputs/day7")
        .expect("Unable to read input")
        .lines()
        .map(|line| line.split(','))
        .flatten()
        .map(|n| n.parse::<i64>().unwrap())
        .collect::<Vec<i64>>();

    let middle = crab_positions.len() / 2;
    let median = *crab_positions.select_nth_unstable(middle).1;

    crab_positions.iter().map(|n| (n - median).abs()).sum()
}

fn cost(n: i64) -> i64 {
    n * (n + 1) / 2
}

pub fn part_2() -> i64 {
    let crab_positions = fs::read_to_string("inputs/day7")
        .expect("Unable to read input")
        .lines()
        .map(|line| line.split(','))
        .flatten()
        .map(|n| n.parse::<i64>().unwrap())
        .collect::<Vec<i64>>();

    (crab_positions.iter().sum::<i64>() / crab_positions.len() as i64..)
        .take(2)
        .map(|t| {
            crab_positions
                .iter()
                .map(|n| cost((n - t).abs()))
                .sum::<i64>()
        })
        .min()
        .unwrap()
}

#[test]
fn test_part_1() {
    assert_eq!(part_1(), 335271);
}

#[test]
fn test_part_2() {
    assert_eq!(part_2(), 95851339);
}

File: 2021/python/day_7.py
import statistics
import math


def part_1():
    with open("inputs/day7", "r") as file:
        crab_positions = list(map(int, file.read().split(",")))
        median = statistics.median(crab_positions)

        return sum(
            map(
                lambda x: abs(x - median),
                crab_positions,
            )
        )


def part_2():
    with open("inputs/day7", "r") as file:
        crab_positions = list(map(int, file.read().split(",")))
        mean = statistics.mean(crab_positions)

        cost = lambda x: x * (x + 1) / 2
        return min(
            sum(
                map(cost, map(abs, map(lambda x: x - math.floor(mean), crab_positions)))
            ),
            sum(
                map(cost, map(abs, map(lambda x: x - math.ceil(mean), crab_positions)))
            ),
        )


print(part_1())
print(part_2())
